### PR TITLE
OAK-10952: improve autogenerated namespace prefixes - test with prototype

### PR DESCRIPTION
**Pull request by reschke**
PoC, hopefully illustrating the idea.

(needs more documentation and support for edge cases, final location in project to be decided)

### CODE REVIEW

**Comment by stefan-egli** [comment]
* could the `makeUpPrefixTryEvenHarder` make a somewhat shorter (perhaps sha-hash) suggestion?
* generally I think this is useful indeed, to avoid later namespace remorse. How impossible would it be to look into support for changing namespaces (I know it's a challenge)

**Comment by reschke** [comment]
> could the makeUpPrefixTryEvenHarder make a somewhat shorter (perhaps sha-hash) suggestion?

SHA-1 would be 160 bits, thus 20 bytes, right? That's not always less than the actual namespace name. I could however special case that case.

> generally I think this is useful indeed, to avoid later namespace remorse. How impossible would it be to look into support for changing namespaces (I know it's a challenge)

That's a separate issue. I believe we already found out that it can be done using the namespace registry, although risky. The point of *this* is to reduce the number of cases where it actually comes up.

**Comment by stefan-egli** [comment]
I was thinking of something like the `cca2a4f` part of a `cca2a4f1dbde1e0f7e337615763ac20a64e39160` git commit

**Comment by reschke** [comment]
> I was thinking of something like the `cca2a4f` part of a `cca2a4f1dbde1e0f7e337615763ac20a64e39160` git commit

-> https://github.com/apache/jackrabbit-oak/pull/2237/commits/7024038f2e79c3069164bf864f2d7652e8c860a1

**Comment by reschke** [comment]
re valid prefixes: the JCR spec allows many characters here; but checking exactly might be hard and overkill.

As valid URIs, strictly speaking, are only ASCII, we should filter out all non-ASCII. That would leaves with a short "allow" list of characters.

**Comment by stefan-egli** [comment]
+1, `"s-1578eb4"` looks nice!

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+public class NamespaceHelperPrototypeTest {
+
+    // map with 'optimal' prefix mappings; hard-wired should be ok
+    private static final Map<String, String> KNOWN_PREFIXES =
+            Map.of("http://purl.org/dc/terms/", "dc",
+                    "http://ns.adobe.com/exif/1.0/", "exif");
+
+    // public API: suggest an available prefix for the provided namespace
+    // (while considering pre-existing mappings)
+    private static @NotNull String suggestPrefix(@NotNull String namespace,
```
Comment (kwin) [inline oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java:45]: suggest is a bit fuzzy to me, why not `getUnusedPrefix(...)`

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    // suggest an available prefix for the provided namespace, based on the characters
+    // in the namespace name
+    // (while considering pre-existing mappings)
+    private static @NotNull String makeUpPrefix(@NotNull String namespace,
+                                                @NotNull Function<String, String> lookupNamespace) {
+        String prefix = namespace.toLowerCase(Locale.ENGLISH);
+
+        // https://www.w3.org/guide/editor/namespaces.html
+        if (prefix.startsWith("http://www.w3c.org/ns/")) {
```
Comment (kwin) [inline oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java:74]: Let us use String constants and their length method to get rid of the magical size.

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    // public API: suggest an available prefix for the provided namespace
+    public static @NotNull String suggestPrefix(String namespace) {
+        return suggestPrefix(namespace, n -> null);
+    }
+
+    // suggest an available prefix for the provided namespace, based on the characters
+    // in the namespace name
+    // (while considering pre-existing mappings)
+    private static @NotNull String makeUpPrefix(@NotNull String namespace,
+                                                @NotNull Function<String, String> lookupNamespace) {
```
Comment (kwin) [inline oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java:70]: `lookupNamespace` is hard to grasp why not `getRegisteredNamespaceForPrefix`

**Review by kwin — COMMENTED** [review]
Overall looks good, some minor remarks only.

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java (hunk 3 above)
Comment (mbaedke) [reply oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java:70]: Shouldn't it be rather named getRegisteredPrefix() here?

Comment (kwin) [reply oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/NamespaceHelperPrototypeTest.java:70]: I think this returns a namespace URI for a given prefix (not the other way around). But this definitely calls for a non ambiguous name.

**Comment by kwin** [comment]
I would place this in [`org.apache.jackrabbit.oak.namepath.impl.GlobalNameMapper`](https://github.com/apache/jackrabbit-oak/blob/trunk/oak-core/src/main/java/org/apache/jackrabbit/oak/namepath/impl/GlobalNameMapper.java) and then call this from according `getOrCreateOakName(orNull)` which should be used from all JCR methods potentially creating items (with expanded names). This should be imho the only place dealing with name resolving in Oak.
